File: clojure/day09.clj
(ns day09
  (:require aoc
            [clojure.data.int-map :refer [dense-int-set]]
            [clojure.math :as math]
            [clojure.string :as str]))


(def ^:const N 500)
(def ^:const start (* (inc N) (/ N 2)))


(defn parse-motion [line]
  (->> (str/split line #" ")
       ((fn [[dir amount]]
          (repeat (parse-long amount) (keyword dir))))))

(defn move-head [^long pt cmd]
  (case cmd
    :U (- pt N)
    :D (+ pt N)
    :L (dec pt)
    :R (inc pt)))

(defn follow ^long [^long head ^long tail]
  (let [hx (rem head N)
        hy (quot head N)
        tx (rem tail N)
        ty (quot tail N)
        dx (- hx tx)
        dy (- hy ty)]
    (if (< (max (abs dx) (abs dy)) 2)
      tail
      (+ tail (long (math/signum dx)) (* N (long (math/signum dy)))))))

(defn move-tail [head tail]
  (reduce
   (fn [rope tail-piece]
     (->> tail-piece
          (follow (peek rope))
          (conj rope)))
   [head]
   tail))

(defn move-rope [[head & tail] cmd]
  (-> head
      (move-head cmd)
      (move-tail tail)))

(defn simulate [motions]
  (reduce
   (fn [[rope seen-2 seen-10] cmd]
     (let [new-pos (move-rope rope cmd)]
       [new-pos
        (conj! seen-2  (nth new-pos 1))
        (conj! seen-10 (nth new-pos 9))]))
   [(vec (repeat 10 start))
    (transient (dense-int-set))
    (transient (dense-int-set))]
   motions))

(defn solve
  ([] (solve 9))
  ([input]
   (let [motions   (mapcat parse-motion (aoc/read-input input))
         [_ p1 p2] (simulate motions)]
     [(count p1)
      (count p2)])))


(solve)

File: python/day09.py
from aoc import *


DIRECTIONS = {'U': -1j, 'D': 1j, 'L': -1, 'R': 1}


def solve(file=9):
    commands = mapl(str.split, read_input(file))
    snake = [0] * 10
    snake_paths = [set() for _ in snake]
    for cmd, amount in commands:
        for _ in range(int(amount)):
            snake[0] += DIRECTIONS[cmd]
            for i, tail in enumerate(snake[1:], 1):
                diff = snake[i-1] - tail
                if abs(diff) >= 2:
                    snake[i] += complex(sign(diff.real), sign(diff.imag))
                snake_paths[i].add(snake[i])
    return len(snake_paths[1]), len(snake_paths[9])


print(solve())










def test():
    assert solve("09_test") == (13, 1)
    assert solve("09_test2") == (88, 36)
    assert solve() == (6197, 2562)
